core/src/main/java/org/bitcoinj/core/Wallet.java

```diff
@@ -2146,12 +2250,42 @@ public boolean maybeCommitTx(Transaction tx) throws VerificationException {
             // move any transactions that are now fully spent to the spent map so we can skip them when creating future
             // spends.
             updateForSpends(tx, false);
-            // Add to the pending pool. It'll be moved out once we receive this transaction on the best chain.
-            // This also registers txConfidenceListener so wallet listeners get informed.
-            log.info("->pending: {}", tx.getHashAsString());
-            tx.getConfidence().setConfidenceType(ConfidenceType.PENDING);
-            confidenceChanged.put(tx, TransactionConfidence.Listener.ChangeReason.TYPE);
-            addWalletTransaction(Pool.PENDING, tx);
+            
+            Set<Transaction> doubleSpendPendingTxns = findDoubleSpendsAgainst(tx, pending);
+            Set<Transaction> doubleSpendUnspentTxns = findDoubleSpendsAgainst(tx, unspent);
+            Set<Transaction> doubleSpendSpentTxns = findDoubleSpendsAgainst(tx, spent);
+            
+            if (!doubleSpendUnspentTxns.isEmpty() || 
+                !doubleSpendSpentTxns.isEmpty() || 
+                !isNotSpendingTxnsInConfidenceType(tx, ConfidenceType.DEAD)) {
+                // tx is a double spend against a tx already in the best chain or spends outputs of a DEAD tx. 
+                // Add tx to the dead pool and schedule confidence listener notifications. 
+                log.info("->dead: {}", tx.getHashAsString());
+                tx.getConfidence().setConfidenceType(ConfidenceType.DEAD);                
+                confidenceChanged.put(tx, TransactionConfidence.Listener.ChangeReason.TYPE);                
+                addWalletTransaction(Pool.DEAD, tx);
+            } else if (!doubleSpendPendingTxns.isEmpty() ||
+                !isNotSpendingTxnsInConfidenceType(tx, ConfidenceType.IN_CONFLICT)) {
+                // tx is a double spend against a pending tx or spends outputs of a tx already IN_CONFLICT
+                // Add tx to the pending pool. Update the confidence type of tx, the txns in conflict with tx and all 
+                // their dependencies to IN_CONFLICT and schedule confidence listener notifications.
+                log.info("->pending: {}", tx.getHashAsString());
```

Comment:
done. Please check you are happy with my change. Also updateForSpends() logs about double spends.